README.md
[![Go Report Card](https://goreportcard.com/badge/github.com/general-error/icepid)](https://goreportcard.com/report/github.com/general-error/icepid)

# Icepid

Icepid is a very simple dashboard. It should be good to use on a small personal
server.

# Features

You can view the following data:

- Uptime, system time
- Available updates
- System logs
- free, w, df
- sensors
- SMART data

# Screenshots

![Login page](screenshots/icepid1.png?raw=true)
![Index page](screenshots/icepid2.png?raw=true)

# Build

Run this in `src` folder:

```sh
go build -o icepid *.go
go build icepid-smart-plugin/*.go
sudo chown root icepid-smart
sudo chmod u+s  icepid-smart
```

# Setup

Drop the icepid & icepid-smart binaries somewhere on your server, edit the
icepid.json settings file and let it run via cron/systemd/etc. `icepid.service`
file is included.

# License

Licensed under terms of GNU General Public License version 3.


CHANGELOG.md
# Change Log

## [1.1.1] - 2017-05-09
### Fixed
- Remove unreachable code
- Use mixed caps naming
- Simplify code
- Run go fmt

### Added
- Go Report Card

## [1.1.0] - 2017-05-09
### Added
- SMART data query
- systemd service unit

### Fixed
- Missing return statement after redirect

### Changed
- Use go library instead of system date command
- Update screenshot

## [1.0.4] - 2017-02-09
### Added
- Error message on wrong password.

## [1.0.3] - 2016-12-31
### Added
- Temperature info using 'sensors'.

### Fixed
- Version div position.

## [1.0.2] - 2016-12-15
### Security
- Make 20s pause after 5 login attempts.

### Added
- This changelog.

## [1.0.1] - 2016-12-08
### Added
- New blue header background.
- Check configuration file version on start.

### Changed
- Screenshots updated

## [1.0.0] - 2016-12-01
### Added
- Initial release.


src/icepid.go
/*
 * Simple server dashboard
 *
 * Copyright 2016 2017 by general-error
 *
 * This file is part of Icepid.
 *
 * Icepid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Icepid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Icepid.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
)

type Settings struct {
	Version      int
	Port         string
	CookieMaxAge int // in seconds
	CookieId     string
	CookieName   string
	Password     string
	Logs         []string
	Disks        []string
}

type IndexPage struct {
	Version string
	Logs    []string
	Updates string
	Uptime  string
	Free    string
	Date    string
	W       string
	Df      string
	Sensors string
	Disks   []string
}

type LoginPage struct {
	ErrorMessage string
}

const appVersion string = "1.1.1"
const confVersion int = 2

var loginAttempts int
var settings = Settings{}
var loginPage = LoginPage{}

func loadSettings() Settings {
	file, _ := os.Open("icepid.json")
	decoder := json.NewDecoder(file)
	configuration := Settings{}
	err := decoder.Decode(&configuration)

	if configuration.Version != confVersion {
		log.Fatal("Error: configuration file version mismatch\n")
	}

	if err == nil {
		return configuration
	}

	log.Fatal("Error: loadSettings\n", err)
	return configuration
}

func main() {
	settings = loadSettings()
	http.HandleFunc("/", loginHandler)
	http.HandleFunc("/index", indexHandler)
	http.HandleFunc("/logout", logoutHandler)
	http.HandleFunc("/log", logHandler)
	http.HandleFunc("/dmesg", dmesgHandler)
	http.HandleFunc("/disk", smartHandler)
	log.Fatal(http.ListenAndServe(settings.Port, nil))
}


src/handlers.go
/*
 * Copyright 2016 2017 by general-error
 *
 * This file is part of Icepid.
 *
 * Icepid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Icepid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Icepid.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

import (
	"fmt"
	"html/template"
	"log"
	"net/http"
	"strconv"
	"time"
)

func checkAccess(w http.ResponseWriter, r *http.Request) {
	_, err := r.Cookie(settings.CookieName)
	if err != nil {
		http.Redirect(w, r, "/", http.StatusFound)
	}
}

func loginHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method == "GET" {
		_, err := r.Cookie(settings.CookieName)
		if err == nil {
			// cookie set
			http.Redirect(w, r, "/index", http.StatusFound)
		} else {
			t, err := template.New("login").Parse(loginTpl)
			if err != nil {
				log.Fatal(err)
			}
			t.Execute(w, loginPage)
		}
	} else {
		if r.FormValue("password") == settings.Password {
			// password right
			loginAttempts = 0
			loginPage.ErrorMessage = ""
			cookie := http.Cookie{
				Name:   settings.CookieName,
				Value:  settings.CookieId,
				MaxAge: settings.CookieMaxAge,
			}
			http.SetCookie(w, &cookie)
			http.Redirect(w, r, "/index", http.StatusFound)
		} else {
			// password wrong
			loginPage.ErrorMessage = "Wrong password!"
			if loginAttempts >= 5 {
				time.Sleep(20 * time.Second)
			} else {
				loginAttempts++
				http.Redirect(w, r, "/", http.StatusFound)
			}
		}
	}
}

func logoutHandler(w http.ResponseWriter, r *http.Request) {
	cookie := http.Cookie{
		Name:   settings.CookieName,
		Value:  settings.CookieId,
		MaxAge: -1,
	}
	http.SetCookie(w, &cookie)
	http.Redirect(w, r, "/", http.StatusFound)
}

func indexHandler(w http.ResponseWriter, r *http.Request) {
	checkAccess(w, r)
	logs := settings.Logs
	updates := getUpdates()
	uptime := getUptime()
	free := getFree()
	date := getDate()
	wVal := getW()
	df := getDf()
	sensors := getSensors()
	disks := settings.Disks

	page := IndexPage{appVersion, logs, updates, uptime, free, date, wVal, df, sensors, disks}

	t, err := template.New("index").Parse(indexTpl)
	if err != nil {
		log.Fatal(err)
	}
	err = t.Execute(w, page)
	if err != nil {
		log.Fatal(err)
	}
}

func logHandler(w http.ResponseWriter, r *http.Request) {
	checkAccess(w, r)

	logNum, err := strconv.Atoi(r.URL.Query().Get("log"))

	if err != nil || logNum < 0 || logNum > len(settings.Logs) {
		http.Redirect(w, r, "/index", http.StatusFound)
		return
	}

	fmt.Fprintf(w, "%v", getLog(settings.Logs[logNum]))
}

func dmesgHandler(w http.ResponseWriter, r *http.Request) {
	checkAccess(w, r)

	fmt.Fprintf(w, "%v", getDmesg())
}

func smartHandler(w http.ResponseWriter, r *http.Request) {
	checkAccess(w, r)

	disk := r.URL.Query().Get("disk")

	if disk != settings.Disks[0] {
		http.Redirect(w, r, "/index", http.StatusFound)
		return
	}

	fmt.Fprintf(w, "%v", getSmart(disk))
}


src/templates.go
/*
 * Copyright 2016 2017 by general-error
 *
 * This file is part of Icepid.
 *
 * Icepid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Icepid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Icepid.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

func wrap(base string) string {
	const header = `<html>
    <head><title>Icepid</title></head>
    <body>
    <style type="text/css">
    .container {
        display: flex;
        align-items: center;
        justify-content: center;
        height: 97vh; /* 100vh causes scrollbar in firefox */
        position: relative;
    }
    .title {
        align-self: flex-start;
        left: 0;
        top: 0;
        position: absolute;
        font-size: 300%;
    }
    .error {
        position: absolute;
        right: 0;
        top: 0;
        padding: 10px;
        font-size: 200%;
        color: red;
    }
    .letter::first-letter {
            color: #0000ff;
            font-size: 200%;
    }
    .header {
        display: flex;
        justify-content: space-between;
        background-color: #b6dcfc;
        margin: -8px;
        padding: 8px;
    }
    .cell {
        border: 1px solid black;
        display: inline-block;
        padding: 3px;
    }
    .bottom {
        position: fixed;
        font-size: small;
        background-color: #b6dcfc;
        bottom: 0;
        right: 0;
        left: 0;
        padding: 3px;
        height: 15px;
    }
    .main {
        margin-bottom: 30px;
    }
    </style>
    `
	const footer = `</body></html>`

	return header + base + footer
}

var loginTpl string = wrap(
	`<div class="container">
        <div class="title">
            <p class="letter">Icepid</p>
        </div>
        <div>
            <form action="/" method="post">
                <input name="password" type="password" autofocus required >
                <input type="submit" value="Login">
            </form>
        </div>
        <div class="error">
            {{.ErrorMessage}}
        </div>
    </div>
    `)

var indexTpl string = wrap(
	`<div class="header">
        <form action="/logout" method="post">
            <input type="submit" value="Logout">
        </form>
        <div>{{.Uptime}}</div>
        <div>{{.Date}}</div>
    </div>
    <div class="main">
        <p>Updates:</p>
        <pre>{{.Updates}}</pre>
        <hr>
        <pre class="cell">{{.W}}</pre><br>
        <pre class="cell">{{.Free}}</pre><br>
        <pre class="cell">{{.Df}}</pre><br>
        <pre class="cell">{{.Sensors}}</pre><br>
        <hr>
        <p>SMART:</p>
        <ul>
            {{range $key, $val := .Disks}}
            <li><a href="/disk?disk={{$val}}">{{$val}}</a></li>
            {{end}}
        </ul>
        <p>Log files:</p>
        <ul>
            <li><a href="/dmesg">dmesg</a></li>
            {{range $key, $val := .Logs}}
            <li><a href="/log?log={{$key}}">{{$val}}</a></li>
            {{end}}
        </ul>
    </div>
    <div class="bottom">v{{.Version}}</div>
     `)


src/icepid-smart-plugin/icepid-smart.go
/*
 * SMART data reader
 *
 * Copyright 2017 by general-error
 *
 * This file is part of Icepid.
 *
 * Icepid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Icepid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Icepid.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/exec"
	"regexp"
)

const appVersion string = "1.0.0"

var reg = regexp.MustCompile(`^[[:alnum:]]+$`)

func main() {
	var v = flag.Bool("v", false, "show program version and exit")
	var s = flag.String("d", "", "device file name like sda")
	flag.Parse()

	if *v {
		fmt.Println("icepid-smart version " + appVersion)
		os.Exit(0)
	}

	if *s == "" {
		flag.PrintDefaults()
		os.Exit(0)
	}

	if !reg.MatchString(*s) {
		log.Print("Error: wrong drive identifier " + *s)
		os.Exit(1)
	}

	out, err := exec.Command("smartctl", "-a", "/dev/"+*s).Output()

	if err != nil {
		log.Print("Error: cannot read SMART data\n", err)
		os.Exit(1)
	}

	fmt.Print(string(out))
}


src/utils.go
/*
 * Copyright 2016 2017 by general-error
 *
 * This file is part of Icepid.
 *
 * Icepid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Icepid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Icepid.  If not, see <http://www.gnu.org/licenses/>.
 */

package main

import (
	"bytes"
	"io/ioutil"
	"log"
	"os/exec"
	"strings"
	"time"
)

func getLog(path string) string {
	file, err := ioutil.ReadFile(path)

	if err != nil {
		log.Fatal(err)
	}

	return string(file)
}

func getUpdates() string {
	cmd := exec.Command("apt", "list", "--upgradable")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Start()
	err = cmd.Wait()

	var output string

	if err == nil {
		output = out.String()
	}

	if strings.Compare(output, "Listing...\n") == 0 {
		return "No updates."
	}

	return output[len("Listing..."):]
}

func getUptime() string {
	out, err := exec.Command("uptime", "-p").Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: get_uptime\n", err)
	return ""
}

func getDmesg() string {
	out, err := exec.Command("dmesg").Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: query_dmesg\n", err)
	return ""
}

func getFree() string {
	out, err := exec.Command("free", "-h").Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: get_free\n", err)
	return ""
}

func getDate() string {
	return time.Now().Format("2006-01-02 15:04:05")
}

func getW() string {
	out, err := exec.Command("w").Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: get_w\n", err)
	return ""
}

func getDf() string {
	out, err := exec.Command("df", "-h", "/", "/home").Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: get_df\n", err)
	return ""
}

func getSensors() string {
	out, err := exec.Command("sensors").Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: get_sensors\n", err)
	return ""
}

func getSmart(disk string) string {
	out, err := exec.Command("./icepid-smart", "-d", disk).Output()

	if err == nil {
		return string(out)
	}

	log.Print("Error: query_smart\n", err)
	return ""
}
